src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/Egress/S3/MultiPartUploadStreamTests.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Tools.Monitor.Egress.S3;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Microsoft.Diagnostics.Monitoring.Tool.UnitTests.Egress.S3
+{
+    public class MultiPartUploadStreamTests
+    {
+
+        private readonly InMemoryS3 _s3 = new InMemoryS3();
+
+        private IEnumerable<byte[]> WithBytesReturned(int totalBytes)
+        {
+            while(totalBytes > 0)            
+            {
+                var length = Random.Shared.Next(1, Math.Min(totalBytes, 1024));
+                var bytes = new byte[length];
+                Random.Shared.NextBytes(bytes);
+                yield return bytes;
+                totalBytes -= length;
+            }
+        }
+
+        [Fact]
+        public async Task ItShouldWriteMultipleParts()
+        {
+            var uploadId = (await _s3.InitiateMultipartUploadAsync("bucket", "key")).UploadId;
+            await using var stream = new MultiPartUploadStream(_s3, "bucket", "key", uploadId, 1024);
+
+            var allBytes = new List<byte>();
+            foreach (var bytes in WithBytesReturned(MultiPartUploadStream.MinimumSize * 3 - 1024))
```

Comment:
Refactor to `const` that are reused so we don't have magic numbers throughout the test:

```cs
const int Part1ExpectedSize = MultiPartUploadStream.MinimumSize;
const int Part2ExpectedSize = MultiPartUploadStream.MinimumSize;
const int Part3ExpectedSize = MultiPartUploadStream.MinimumSize - 1024;
```